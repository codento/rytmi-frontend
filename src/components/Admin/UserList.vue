<template>
  <div>
    <b-form-group
      label="Filter by name"
    >
      <b-form-input v-model="filterValue" />
    </b-form-group>
    <div>
      Active User<span class="cui-check" />
      Admin User<span class="cui-cog" />
    </div>
    <b-list-group
      v-for="(user, idx) in usersAsList"
      :key="idx"
    >
      <b-list-group-item
        :active="isUserSelected(user.id)"
        name="select-user"
        @click="selectUser(user.id)"
      >
        {{ user.firstName + ' ' + user.lastName }}
        <span
          v-if="user.active"
          class="cui-check"
        />
        <span
          v-if="user.admin"
          class="cui-cog"
        />
      </b-list-group-item>
    </b-list-group>
  </div>
</template>
<script>
import { sortBy } from 'lodash'

export default {
  props: {
    users: {
      type: Object,
      required: true
    }
  },
  data () {
    return {
      filterValue: '',
      selectedUserId: null
    }
  },
  computed: {
    usersAsList () {
      const userList = Object.keys(this.users).map(id => this.users[id]).filter(user => {
        const lowerCase = (user.firstName + ' ' + user.lastName).toLowerCase()
        return lowerCase.includes(this.filterValue.toLowerCase())
      })
      return sortBy(userList, ['firstName', 'lastName'])
    }
  },
  methods: {
    selectUser (id) {
      this.selectedUserId = id
      this.$emit('user-selected', id)
    },
    isUserSelected (id) {
      return this.selectedUserId === id
    }
  }
}
</script>

<style lang="scss" scoped>
  .cui-check {
    color: green
  }
</style>
