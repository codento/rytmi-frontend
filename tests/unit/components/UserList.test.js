import Vuex from 'vuex'
import BootstrapVue from 'bootstrap-vue'
import { createLocalVue, mount } from '@vue/test-utils'
import UserList from '@/components/Admin/UserList'

describe('UserList.vue', () => {
  const localVue = createLocalVue()
  localVue.use(Vuex)
  localVue.use(BootstrapVue)

  const propsData = {
    users: {
      1: {
        id: 1,
        firstName: 'Aaa',
        lastName: 'Foo',
        active: true,
        admin: true
      },
      2: {
        id: 2,
        firstName: 'Bbb',
        lastName: 'Rab',
        active: true,
        admin: true
      },
      3: {
        id: 3,
        firstName: 'Ccc',
        lastName: 'Foo',
        active: false,
        admin: true
      },
      4: {
        id: 4,
        firstName: 'Ddd',
        lastName: 'Baru',
        active: true,
        admin: false
      }
    }
  }

  let wrapper

  beforeEach(() => {
    wrapper = mount(UserList, {
      propsData,
      localVue
    })
  })

  it('List is rendered with correct length', () => {
    const list = wrapper.findAll('.list-group-item')
    expect(list).toHaveLength(4)
  })

  it('should select user', () => {
    const selectUser = jest.spyOn(wrapper.vm, 'selectUser')
    const list = wrapper.findAll('.list-group-item')
    list.at(2).trigger('click')
    expect(wrapper.vm.selectedUserId).toBe(3)
    expect(selectUser).toHaveBeenCalled()
    expect(list.at(2).classes()).toContain('active')
  })

  it('Template is correct', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
