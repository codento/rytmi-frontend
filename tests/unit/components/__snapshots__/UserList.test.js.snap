// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList.vue Template is correct 1`] = `
<div>
  <fieldset
    class="form-group"
    id="__BVID__12"
  >
    <legend
      class="col-form-label pt-0"
      id="__BVID__12__BV_label_"
      tabindex="-1"
    >
      Filter by name
    </legend>
    <div
      role="group"
      tabindex="-1"
    >
      <input
        class="form-control"
        id="__BVID__13"
        type="text"
      />
      <!---->
      <!---->
      <!---->
    </div>
  </fieldset>
   
  <div>
    
    Active User
    <span
      class="cui-check"
    />
    
    Admin User
    <span
      class="cui-cog"
    />
  </div>
   
  <div
    class="list-group"
  >
    <div
      class="list-group-item"
    >
      
      Aaa Foo
      
      <span
        class="cui-check"
      />
       
      <span
        class="cui-cog"
      />
    </div>
  </div>
  <div
    class="list-group"
  >
    <div
      class="list-group-item"
    >
      
      Bbb Rab
      
      <span
        class="cui-check"
      />
       
      <span
        class="cui-cog"
      />
    </div>
  </div>
  <div
    class="list-group"
  >
    <div
      class="list-group-item"
    >
      
      Ccc Foo
      
      <!---->
       
      <span
        class="cui-cog"
      />
    </div>
  </div>
  <div
    class="list-group"
  >
    <div
      class="list-group-item"
    >
      
      Ddd Baru
      
      <span
        class="cui-check"
      />
       
      <!---->
    </div>
  </div>
</div>
`;
